# Naive Bayes

In [1]:
import os
import io
import numpy
from pandas import DataFrame
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB


def readFiles(path):
    for root, dirnames, filenames in os.walk(path):
        for filename in filenames:
            path = os.path.join(root, filename)

            inBody = False
            lines = []
            f = io.open(path, 'r', encoding='latin1')
            for line in f:
                if inBody:
                    lines.append(line)
                elif line == '\n':
                    inBody = True
            f.close()
            message = '\n'.join(lines)
            yield path, message


def dataFrameFromDirectory(path, classification):
    rows = []
    index = []
    for filename, message in readFiles(path):
        rows.append({'message': message, 'class': classification})
        index.append(filename)

    return DataFrame(rows, index=index)

data = DataFrame({'message': [], 'class': []})

data = data.append(dataFrameFromDirectory("C:\\Users\\syed mubarak pasha\\OneDrive\\Documents\\Python\\Assignment\\Naive Bayes\\spam", 'spam'))
data = data.append(dataFrameFromDirectory("C:\\Users\\syed mubarak pasha\\OneDrive\\Documents\\Python\\Assignment\\Naive Bayes\\ham", 'ham'))


C:\Users\syed mubarak pasha\AppData\Local\Temp\ipykernel_16304\2741527179.py:38: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  data = data.append(dataFrameFromDirectory("C:\\Users\\syed mubarak pasha\\OneDrive\\Documents\\Python\\Assignment\\Naive Bayes\\spam", 'spam'))
C:\Users\syed mubarak pasha\AppData\Local\Temp\ipykernel_16304\2741527179.py:39: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  data = data.append(dataFrameFromDirectory("C:\\Users\\syed mubarak pasha\\OneDrive\\Documents\\Python\\Assignment\\Naive Bayes\\ham", 'ham'))


In [2]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2983 entries, C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\spam\00001.7848dde101aa985090474a91ec93fcf0 to C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\ham\02500.05b3496ce7bca306bed0805425ec8621
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   message  2983 non-null   object
 1   class    2983 non-null   object
dtypes: object(2)
memory usage: 69.9+ KB


In [3]:
data.head()

message  \
C:\Users\syed mubarak pasha\OneDrive\Documents\...  <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Tr...   
C:\Users\syed mubarak pasha\OneDrive\Documents\...  1) Fight The Risk of Cancer!\n\nhttp://www.adc...   
C:\Users\syed mubarak pasha\OneDrive\Documents\...  1) Fight The Risk of Cancer!\n\nhttp://www.adc...   
C:\Users\syed mubarak pasha\OneDrive\Documents\...  ##############################################...   
C:\Users\syed mubarak pasha\OneDrive\Documents\...  I thought you might like these:\n\n1) Slim Dow...   

                                                   class  
C:\Users\syed mubarak pasha\OneDrive\Documents\...  spam  
C:\Users\syed mubarak pasha\OneDrive\Documents\...  spam  
C:\Users\syed mubarak pasha\OneDrive\Documents\...  spam  
C:\Users\syed mubarak pasha\OneDrive\Documents\...  spam  
C:\Users\syed mubarak pasha\OneDrive\Documents\...  spam

In [4]:
data = data.replace(['ham','spam'],[0, 1])
data.head(10)

message  \
C:\Users\syed mubarak pasha\OneDrive\Documents\...  <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Tr...   
C:\Users\syed mubarak pasha\OneDrive\Documents\...  1) Fight The Risk of Cancer!\n\nhttp://www.adc...   
C:\Users\syed mubarak pasha\OneDrive\Documents\...  1) Fight The Risk of Cancer!\n\nhttp://www.adc...   
C:\Users\syed mubarak pasha\OneDrive\Documents\...  ##############################################...   
C:\Users\syed mubarak pasha\OneDrive\Documents\...  I thought you might like these:\n\n1) Slim Dow...   
C:\Users\syed mubarak pasha\OneDrive\Documents\...  A POWERHOUSE GIFTING PROGRAM You Don't Want To...   
C:\Users\syed mubarak pasha\OneDrive\Documents\...  Help wanted.  We are a 14 year old fortune 500...   
C:\Users\syed mubarak pasha\OneDrive\Documents\...  <html>\n\n<head>\n\n<title>ReliaQuote - Save U...   
C:\Users\syed mubarak pasha\OneDrive\Documents\...  TIRED OF THE BULL OUT THERE?\n\nWant To Stop L...   
C:\Users\syed mubarak pasha\OneDrive\Documents\...  Dear ricardo1 ,\n\n\n\n<html>\n\n<body>\n\n<ce...   

                                                    class  
C:\Users\syed mubarak pasha\OneDrive\Documents\...      1  
C:\Users\syed mubarak pasha\OneDrive\Documents\...      1  
C:\Users\syed mubarak pasha\OneDrive\Documents\...      1  
C:\Users\syed mubarak pasha\OneDrive\Documents\...      1  
C:\Users\syed mubarak pasha\OneDrive\Documents\...      1  
C:\Users\syed mubarak pasha\OneDrive\Documents\...      1  
C:\Users\syed mubarak pasha\OneDrive\Documents\...      1  
C:\Users\syed mubarak pasha\OneDrive\Documents\...      1  
C:\Users\syed mubarak pasha\OneDrive\Documents\...      1  
C:\Users\syed mubarak pasha\OneDrive\Documents\...      1

In [5]:
data["message"][:10]

C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\spam\00001.7848dde101aa985090474a91ec93fcf0    <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Tr...
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\spam\00002.d94f1b97e48ed3b553b3508d116e6a09    1) Fight The Risk of Cancer!\n\nhttp://www.adc...
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\spam\00003.2ee33bc6eacdb11f38d052c44819ba6c    1) Fight The Risk of Cancer!\n\nhttp://www.adc...
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\spam\00004.eac8de8d759b7e74154f142194282724    ##############################################...
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\spam\00005.57696a39d7d84318ce497886896bf90d    I thought you might like these:\n\n1) Slim Dow...
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\spam\00006.5ab5620d3d7c6c0db76234556a16f6c1    A POWE

In [6]:
data["class"][10:]

C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\spam\00011.61816b9ad167657773a427d890d0468e    1
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\spam\00012.381e4f512915109ba1e0853a7a8407b2    1
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\spam\00013.d3f0b591a65f116ea5d9d4ad919f83aa    1
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\spam\00014.7d38c46424f24fc8012ac15a95a2ac14    1
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\spam\00015.048434ab64c86cf890eda1326a5643f5    1
                                                                                                                           ..
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\ham\02496.aae0c81581895acfe65323f344340856     0
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\ham\02497.60497db0a06c2132ec2374b2898084d

In [7]:
data['class'].value_counts()

0    2483
1     500
Name: class, dtype: int64

In [8]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(data['message'], data['class'], test_size = 0.2, train_size = 0.8, random_state = 1)

In [9]:
print('test set: ' + str(x_test.shape))


test set: (597,)


In [10]:
print('train set:' + str(x_train.shape))

train set:(2386,)


So we have 597 messages for testing, and 2386 messages for training!

In [11]:
print(x_train)

C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\ham\00301.48ccf486575754a29b80e4eae2c5e227     > From: fork-admin@xent.com [mailto:fork-admin...
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\ham\01302.06ac4ee24e3c434afa330b8c0408649b     Once upon a time, Jon wrote :\n\n\n\n> Since l...
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\ham\01447.98e4b20ceb192594e992f7db9f8dfc53     On Fri, Sep 06, 2002 at 06:22:52PM +0100, Matt...
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\ham\00128.0e92ec0c8bd8233f7e7873e93df43277     what takes time and money: noting the amounts ...
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\spam\00481.5c95b526e965fa325044123c4ce29c1f     =?GB2312?B?yNW12squ0ru97NbQufq5+rzKtefX08nosb...
                                                                                                                                  

In [12]:
print(x_train, y_train)

C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\ham\00301.48ccf486575754a29b80e4eae2c5e227     > From: fork-admin@xent.com [mailto:fork-admin...
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\ham\01302.06ac4ee24e3c434afa330b8c0408649b     Once upon a time, Jon wrote :\n\n\n\n> Since l...
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\ham\01447.98e4b20ceb192594e992f7db9f8dfc53     On Fri, Sep 06, 2002 at 06:22:52PM +0100, Matt...
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\ham\00128.0e92ec0c8bd8233f7e7873e93df43277     what takes time and money: noting the amounts ...
C:\Users\syed mubarak pasha\OneDrive\Documents\Python\Assignment\Naive Bayes\spam\00481.5c95b526e965fa325044123c4ce29c1f     =?GB2312?B?yNW12squ0ru97NbQufq5+rzKtefX08nosb...
                                                                                                                                  

In [13]:
vectorizer = CountVectorizer()
counts = vectorizer.fit_transform(x_train)

classifier = MultinomialNB()
targets = y_train
classifier.fit(counts, targets)

MultinomialNB()

In [14]:
examples = ["$$DON'T WAIT!!! CALL NOW $$ 1-800-896-6568", "Hi Bob, how about a game of golf tomorrow?"]
example_counts = vectorizer.transform(examples)
predictions = classifier.predict(example_counts)
predictions

array([1, 0], dtype=int64)

In [15]:
examples = x_test
example_counts = vectorizer.transform(examples)
predictions_1 = classifier.predict(example_counts)
predictions_1

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0,

In [16]:
features_test_transformed  = vectorizer.transform(x_test)
print("classifier accuracy {:.2f}%".format(classifier.score(features_test_transformed, y_test) * 100))

classifier accuracy 96.98%
